vllm/model_executor/models/bigdl_llama.py

```diff
@@ -0,0 +1,140 @@
+import torch
+from torch import nn
+
+from transformers import AutoTokenizer, AutoModelForCausalLM, PreTrainedTokenizerBase, LlamaConfig
+from typing import Optional, Tuple, List, Type, Dict
+
+from vllm.transformers_utils.tokenizer import (detokenize_incrementally,
+                                               get_tokenizer)
+from vllm.model_executor.quantization_utils import QuantizationConfig
+from vllm.sequence import SamplerOutput, SequenceOutputs
+import math
+
+import pdb
+
+from transformers.generation.logits_process import (
+    LogitsProcessorList,
+    RepetitionPenaltyLogitsProcessor,
+    TemperatureLogitsWarper,
+    TopKLogitsWarper,
+    TopPLogitsWarper,
+)
+
+def prepare_logits_processor(
+    temperature: float, repetition_penalty: float, top_p: float, top_k: int
+) -> LogitsProcessorList:
+    processor_list = LogitsProcessorList()
+    # TemperatureLogitsWarper doesn't accept 0.0, 1.0 makes it a no-op so we skip two cases.
+    if temperature >= 1e-5 and temperature != 1.0:
+        processor_list.append(TemperatureLogitsWarper(temperature))
+    # if repetition_penalty > 1.0:
+    #     processor_list.append(RepetitionPenaltyLogitsProcessor(repetition_penalty))
+    if 1e-8 <= top_p < 1.0:
+        processor_list.append(TopPLogitsWarper(top_p))
+    if top_k > 0:
+        processor_list.append(TopKLogitsWarper(top_k))
+    return processor_list
+
+class BigDLLlamaForCausalLM(nn.Module):
+    def __init__(
+        self,
+        config: LlamaConfig,
+        quant_config: Optional[QuantizationConfig] = None,
+    ):
+        super().__init__()
+        # pdb.set_trace()
+        self.config = config
+        self.model = AutoModelForCausalLM.from_pretrained(config._name_or_path)
+        self.tokenizer = AutoTokenizer.from_pretrained(config._name_or_path)
+
+    def decode(self, generated_ids: List[int]) -> str:
+        return self.tokenizer.decode(
+            generated_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False
+        )
+
+    def forward(
+        self, seq_group_meta_data_lists, kv_cache: Optional = None
+    ) -> Tuple[torch.Tensor, List[Tuple[torch.Tensor, torch.Tensor]]]:
+        kv_cache_0 = self.model.config.num_hidden_layers
+        kv_cache_1 = 2
+        bigdl_kv_cache = [[torch.Tensor() for _ in range(kv_cache_1)] for _ in range(kv_cache_0)]
+        seq_len = len(seq_group_meta_data_lists)
+        for i in range(seq_len):
+            if kv_cache.get(i) is None:
+                kv_cache[i] = bigdl_kv_cache[:]
```

Comment:
If we want to allocate space for it, we should ensure it is a deep copy.

I guess the shape is:

`[[Tensor, Tensor], ... [Tensor, Tensor]]` there are `n_layers` items in total.

Each tensor should be the shape `[batch_size, n_head, seq_len, head_dim]`?